base/inference.jl

```diff
@@ -34,10 +34,18 @@ typealias VarTable Array{Any,1}
 type VarState
     typ
     undef::Bool
+    function VarState(typ::ANY, undef::Bool)
+        if isa(typ, DataType) && isdefined(typ, :instance)
```

Comment:
I see. In that case it would make more sense to do the transformation at the point where the arguments are received. `VarState` is wrapper-like, so it's confusing if it changes the thing you're wrapping in addition to wrapping it.